[TA2577] Updating ZAP (degraded_io_seq) at every 5 seconds if zvol is degraded

mynktl opened the pull request:
In cStor, zrepl shares its checkpointed_ionum to istgt during initial handshake.
Currently, in-memory checkpointed_ionum is not initialized to the value stored in ZAP (and initialized to 0), which makes zrepl to store '0' in ZAP in one of the iteration.

zrepl stores the running_ionum to ZAP only in its healthy mode. If istgt restarts within the time zrepl becomes healthy, istgt gets the same ionum stored in replicas everytime, and thus, making to use same ionum as starting number in istgt.

Changes : 
- Updating ZAP (key: `degraded_io_seq`) at every 5 seconds if ZVOL is in degraded mode, so that, this will be shared to target during handshake
- Initialization of checkpointed_ionum by reading values from ZAP

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

Commit:
Updating ZAP (key : degraded_io_seq) at every 5 seconds if zvol is in degraded mode

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

vishnuitta reviewed (commented):
there is lot of code repetition

Commit:
- Passing ZAP key to API uzfs_zvol_get_last_committed_io_no/uzfs_zvol_store_last_committed_io_no API
- Adding degraded_checkpointed_time in zvol_info_t to keep track of updating io sequence number in degraded mode

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

vishnuitta commented on lib/libzrepl/data_conn.c:
```diff
@@ -48,6 +48,13 @@ uint16_t rebuild_io_server_port = REBUILD_IO_SERVER_PORT;
 kcondvar_t timer_cv;
 kmutex_t timer_mtx;
 
+typedef struct singly_node_list_s {
```
lets move this to zrepl_mgmt.h

mynktl replied:
ok

vishnuitta commented on lib/libzrepl/data_conn.c:
```diff
@@ -591,19 +599,38 @@ uzfs_zvol_timer_thread(void)
 	zvol_info_t *zinfo;
 	time_t min_interval;
 	time_t now, next_check;
+	struct singly_node_list n_zvol_list, n_free_list;
```
lets just make it as zvol_node_list, free_node_list kind of names

mynktl replied:
ok

Commit:
- Addressing review comments
- Initializing zinfo's checkpointed io sequence number with old value stored in ZAP

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

vishnuitta commented on include/zrepl_mgmt.h:
```diff
@@ -180,9 +186,9 @@ extern int uzfs_zinfo_init(void *zv, const char *ds_name,
+uint64_t uzfs_zvol_get_last_committed_io_no(zvol_state_t *zv, char *key);
 void uzfs_zvol_store_last_committed_io_no(zvol_state_t *zv,
-    uint64_t io_seq);
+    uint64_t io_seq, char *key);
```
lets make this as (key, value) rather than (val, key)

mynktl replied:
ok. I will try to cover this change in the next PR.

vishnuitta commented on lib/libzrepl/data_conn.c:
```diff
@@ -613,25 +640,65 @@ uzfs_zvol_timer_thread(void)
+				}
 			}
+
+			if (next_check > now &&
```
why this check? I understand that - next_check should be greater than now, and the code flow guarantees that.
probably, lets add a print if next_check < now.

mynktl replied:
`next_check` and `now` are `unsigned long` so if `next_check` is `< now` then value of `next_check - now` will be very big number!

vishnuitta reviewed (commented):
changes looks good

vishnuitta reviewed (approved):
changes are good
